src/shogun/base/AnyParameter.h

```diff
@@ -21,49 +27,102 @@ namespace shogun
 		GRADIENT_NOT_AVAILABLE = 0,
 		GRADIENT_AVAILABLE = 1
 	};
-
+	/** @brief Class AnyParameterProperties keeps track of parameter properties.
+	 *
+	 * Currently there are three types of parameters:
+	 * -# HYPERPARAMETER: the parameter determines how the training is
+	 * performed, e.g. regularisation
+	 * -# GRADIENT_PARAM: the parameter is used in gradient updates
+	 * -# MODEL_PARAM: model parameter that is trained, e.g. weight and bias
+	 * vectors and bias
+	 */
 	class AnyParameterProperties
 	{
 	public:
-		AnyParameterProperties()
-		    : m_description(), m_model_selection(MS_NOT_AVAILABLE),
-		      m_gradient(GRADIENT_NOT_AVAILABLE)
+		static const int32_t HYPERPARAMETER = 0x00000001;
+		static const int32_t GRADIENT_PARAM = 0x00000010;
+		static const int32_t MODEL_PARAM = 0x00000100;
+
+		/** default constructor where all parameter types are switched to false
+		 */
+		AnyParameterProperties() : m_description(), m_mask_attribute(0x00000000)
 		{
 		}
+		/** legacy constructor */
 		AnyParameterProperties(
 		    std::string description,
 		    EModelSelectionAvailability model_selection = MS_NOT_AVAILABLE,
-		    EGradientAvailability gradient = GRADIENT_NOT_AVAILABLE)
+		    EGradientAvailability gradient = GRADIENT_NOT_AVAILABLE,
+		    bool hyperparameter = false)
 		    : m_description(description), m_model_selection(model_selection),
 		      m_gradient(gradient)
 		{
+			m_mask_attribute = 0x00000000;
+			if (model_selection)
+				m_mask_attribute |= MODEL_PARAM;
+			if (gradient)
+				m_mask_attribute |= GRADIENT_PARAM;
+			if (hyperparameter)
+				m_mask_attribute |= HYPERPARAMETER;
+		}
+		/** mask constructor */
+		AnyParameterProperties(std::string description, int32_t mask_attribute)
+		    : m_description(description)
+		{
+			m_mask_attribute = mask_attribute;
 		}
+		/** copy contructor */
 		AnyParameterProperties(const AnyParameterProperties& other)
 		    : m_description(other.m_description),
 		      m_model_selection(other.m_model_selection),
-		      m_gradient(other.m_gradient)
+		      m_gradient(other.m_gradient),
+		      m_mask_attribute(other.m_mask_attribute)
 		{
 		}
-
+		/** description getter */
 		std::string get_description() const
 		{
 			return m_description;
 		}
-
+		/** model selection flag getter */
 		EModelSelectionAvailability get_model_selection() const
 		{
-			return m_model_selection;
+			EModelSelectionAvailability return_val;
+			if (m_mask_attribute & HYPERPARAMETER)
+				return_val = EModelSelectionAvailability::MS_AVAILABLE;
+			else
+				return_val = EModelSelectionAvailability::MS_NOT_AVAILABLE;
+			return return_val;
 		}
 
+		/** gradient flag getter */
 		EGradientAvailability get_gradient() const
 		{
-			return m_gradient;
+			EGradientAvailability return_val;
+			if (m_mask_attribute & GRADIENT_PARAM)
+				return_val = EGradientAvailability::GRADIENT_AVAILABLE;
```

Comment:
it makes sense to return immediately without return_val

Reply:
Do you mean with a one liner like this? 
```
return static_cast<EGradientAvailability>((m_mask_attribute & GRADIENT_PARAM) > 0);
```

Reply:
Yeah maybe with some helper function so it looks shorter? There is some repeated code in these flags.